src/ncp/posix/netif.cpp

```diff
@@ -114,6 +114,65 @@ void Netif::UpdateIp6UnicastAddresses(const std::vector<Ip6AddressInfo> &aAddrIn
     mIp6UnicastAddresses.assign(aAddrInfos.begin(), aAddrInfos.end());
 }
 
+otbrError Netif::UpdateIp6MulticastAddresses(const std::vector<Ip6Address> &aAddrs)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    // Remove stale addresses
+    for (const Ip6Address &address : mIp6MulticastAddresses)
+    {
+        if (std::find(aAddrs.begin(), aAddrs.end(), address) == aAddrs.end())
+        {
+            otbrLogInfo("Remove address: %s", Ip6Address(address).ToString().c_str());
+            SuccessOrExit(error = ProcessMulticastAddressChange(address, /* aIsAdded */ false));
+        }
+    }
+
+    // Add new addresses
+    for (const Ip6Address &address : aAddrs)
+    {
+        if (std::find(mIp6MulticastAddresses.begin(), mIp6MulticastAddresses.end(), address) ==
+            mIp6MulticastAddresses.end())
+        {
+            otbrLogInfo("Add address: %s", Ip6Address(address).ToString().c_str());
+            SuccessOrExit(error = ProcessMulticastAddressChange(address, /* aIsAdded */ true));
+        }
+    }
+
+    mIp6MulticastAddresses.assign(aAddrs.begin(), aAddrs.end());
+
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        mIp6MulticastAddresses.clear();
+    }
+    return error;
+}
+
+otbrError Netif::ProcessMulticastAddressChange(const Ip6Address &aAddress, bool aIsAdded)
+{
+    struct ipv6_mreq mreq;
+    otbrError        error = OTBR_ERROR_NONE;
+    int              err;
+
+    VerifyOrExit(mIpFd >= 0);
```

Comment:
mIpFd < 0 seems to be a serious problem, shall we return an error for it or fail harder?

Reply:
Returned 'INVALID_STATE' for it.